File: raytracer_rust_common/src/camera.rs
use crate::{
    components::{
        material::Material,
        surface::{Ray, Surface},
    },
    util::{
        interval::{Interval, interval},
        vec3::{Arrow, Color, Point, arrow, color, point},
    },
};

#[derive(Debug, Clone, Copy)]
pub struct CameraRenderOptions {
    pub image_width: usize,
    pub aspect_ratio: f64,
    pub max_depth: usize,
}

impl Default for CameraRenderOptions {
    fn default() -> Self {
        Self {
            image_width: 400,
            aspect_ratio: 16.0 / 9.0,
            max_depth: 10,
        }
    }
}

#[derive(Debug, Clone, Copy)]
pub enum Background {
    Sky,
    Solid(Color),
}

#[derive(Debug, Clone, Copy)]
pub struct CameraSceneOptions {
    pub vertical_fov: f64,
    pub look_from: Point,
    pub look_at: Point,
    pub v_up: Arrow,
    pub defocus_angle: f64,
    pub focus_distance: f64,
    pub background: Background,
}

impl Default for CameraSceneOptions {
    fn default() -> Self {
        Self {
            vertical_fov: 90.0,
            look_from: point(0.0, 0.0, 0.0),
            look_at: point(0.0, 0.0, -1.0),
            v_up: arrow(0.0, 1.0, 0.0),
            defocus_angle: 0.0,
            focus_distance: 10.0,
            background: Background::Sky,
        }
    }
}

struct CameraComputed {
    image_height: usize,
    first_pixel_location: Point,
    pixel_delta_u: Arrow,
    pixel_delta_v: Arrow,
    defocus_disk_u: Arrow,
    defocus_disk_v: Arrow,
}

#[allow(dead_code)]
pub struct Camera {
    render: CameraRenderOptions,
    scene: CameraSceneOptions,
    computed: CameraComputed,
    pixel_color: Vec<Color>,
}

impl Camera {
    pub fn new(render: CameraRenderOptions, scene: CameraSceneOptions) -> Self {
        let image_height = (render.image_width as f64 / render.aspect_ratio).max(1.0) as usize;

        // Viewport dimensions
        let theta = scene.vertical_fov.to_radians();
        let h = (theta / 2.0).tan();
        let viewport_height = 2.0 * h * scene.focus_distance;
        let viewport_width = viewport_height * (render.image_width as f64 / image_height as f64);

        // Camera basis vectors
        let w = (scene.look_from - scene.look_at).unit_vector().as_arrow();
        let u = scene.v_up.cross(w).unit_vector();
        let v = w.cross(u);

        // Pixel basis vectors
        let viewport_u = viewport_width * u;
        let viewport_v = viewport_height * -v;

        let pixel_delta_u = viewport_u / render.image_width as f64;
        let pixel_delta_v = viewport_v / image_height as f64;

        let viewport_upper_left =
            scene.look_from - (scene.focus_distance * w) - viewport_u / 2.0 - viewport_v / 2.0;
        let first_pixel_location = viewport_upper_left + 0.5 * (pixel_delta_u + pixel_delta_v);

        // Defocus disk basis vectors
        let defocus_radius = scene.focus_distance * (scene.defocus_angle / 2.0).to_radians().tan();
        let defocus_disk_u = u * defocus_radius;
        let defocus_disk_v = v * defocus_radius;

        let computed = CameraComputed {
            image_height,
            first_pixel_location,
            pixel_delta_u,
            pixel_delta_v,
            defocus_disk_u,
            defocus_disk_v,
        };

        let pixel_color = vec![color(0.0, 0.0, 0.0); render.image_width * image_height];

        Camera {
            render,
            scene,
            computed,
            pixel_color,
        }
    }

    pub fn render_pixel(&mut self, i: usize, j: usize, n: usize, world: &impl Surface) -> [u8; 3] {
        let index = j * self.render.image_width + i;

        let pixel_color = self.pixel_color[index];

        let ray = self.get_ray(i as f64, j as f64);
        let color = self.get_color(ray, self.render.max_depth, world);

        let pixel_color = pixel_color + color;
        self.pixel_color[index] = pixel_color;

        self.convert_color(pixel_color / n as f64)
    }

    fn get_color(&self, ray: Ray, depth: usize, world: &impl Surface) -> Color {
        if depth == 0 {
            return color(0.0, 0.0, 0.0);
        }

        let Some(hit) = world.hit(ray, interval(0.001, f64::INFINITY)) else {
            return match self.scene.background {
                Background::Solid(color) => color,
                Background::Sky => {
                    let unit_direction = ray.direction.unit_vector();
                    let a = 0.5 * (unit_direction.y() + 1.0);
                    color(1.0, 1.0, 1.0) * (1.0 - a) + color(0.5, 0.7, 1.0) * a
                }
            };
        };

        let emission_color = hit.material.emitted(&hit);

        let Some(mat_hit) = hit.material.scatter(ray, hit) else {
            return emission_color;
        };

        let scatter_color =
            mat_hit.attenuation * self.get_color(mat_hit.scattered, depth - 1, world);

        emission_color + scatter_color
    }

    fn get_ray(&self, i: f64, j: f64) -> Ray {
        let offset = arrow(
            Interval::UNIT.random_double() - 0.5,
            Interval::UNIT.random_double() - 0.5,
            0.0,
        );
        let pixel_sample = self.computed.first_pixel_location
            + ((i + offset.x()) * self.computed.pixel_delta_u)
            + ((j + offset.y()) * self.computed.pixel_delta_v);

        let origin = if self.scene.defocus_angle <= 0.0 {
            self.scene.look_from
        } else {
            let p = Point::random_in_unit_disk();
            self.scene.look_from
                + (p.x() * self.computed.defocus_disk_u)
                + (p.y() * self.computed.defocus_disk_v)
        };
        let direction = (pixel_sample - origin).as_arrow();
        let time = Interval::UNIT.random_double();

        Ray::new(origin, direction, time)
    }

    fn convert_color(&self, color: Color) -> [u8; 3] {
        let intensity = interval(0.000, 0.999);
        let color = color.map(|a| {
            if a > 0.0 {
                256.0 * intensity.clamp(a.sqrt())
            } else {
                0.0
            }
        });

        [color.r() as _, color.g() as _, color.b() as _]
    }

    pub fn image_width(&self) -> usize {
        self.render.image_width
    }

    pub fn image_height(&self) -> usize {
        self.computed.image_height
    }
}

File: raytracer_ts/camera.ts
import { Ray, type Hittable } from './hittable';
import { degreesToRadians, Interval, Vec3, type Color3, type Point3 } from './util';

export interface CameraRenderOptions {
    imageWidth: number;
    aspectRatio: number;
    samplesPerPixel: number;
    maxDepth: number;
}

export interface CameraSceneOptions {
    verticalFov: number;
    lookFrom: Vec3;
    lookAt: Vec3;
    vUp: Vec3;

    defocusAngle: number;
    focusDistance: number;

    background?: Color3;
}

export class Camera {
    public imageWidth: number;
    public imageHeight: number;

    private center: Point3;
    private upperLeftPixelLocation: Point3;
    private pixelDeltaU: Vec3;
    private pixelDeltaV: Vec3;

    private u: Vec3;
    private v: Vec3;
    private w: Vec3;
    private defocusDiskU: Vec3;
    private defocusDiskV: Vec3;

    constructor(
        private renderOptions: CameraRenderOptions,
        private sceneOptions: CameraSceneOptions,
    ) {
        this.imageWidth = renderOptions.imageWidth;
        this.imageHeight = Math.max(1, Math.trunc(this.imageWidth / renderOptions.aspectRatio));

        this.center = sceneOptions.lookFrom;

        const viewportHeight =
            2.0 * sceneOptions.focusDistance * Math.tan(degreesToRadians(sceneOptions.verticalFov) / 2);
        const viewportWidth = viewportHeight * (this.imageWidth / this.imageHeight);

        this.w = sceneOptions.lookFrom.minus(sceneOptions.lookAt).unitVector;
        this.u = sceneOptions.vUp.cross(this.w).unitVector;
        this.v = this.w.cross(this.u);

        const viewportU = this.u.times(viewportWidth);
        const viewportV = this.v.times(-viewportHeight);

        this.pixelDeltaU = viewportU.div(this.imageWidth);
        this.pixelDeltaV = viewportV.div(this.imageHeight);

        const viewportUpperLeft = this.center
            .minus(this.w.times(sceneOptions.focusDistance))
            .minus(viewportU.div(2))
            .minus(viewportV.div(2));

        this.upperLeftPixelLocation = viewportUpperLeft
            .plus(this.pixelDeltaU.times(0.5))
            .plus(this.pixelDeltaV.times(0.5));

        const defocusRadius = sceneOptions.focusDistance * Math.tan(degreesToRadians(sceneOptions.defocusAngle / 2));
        this.defocusDiskU = this.u.times(defocusRadius);
        this.defocusDiskV = this.v.times(defocusRadius);
    }

    public async render(world: Hittable, data: ImageDataArray) {
        const BATCH_THRESHOLD_MS = 50;
        let lastYieldTime = performance.now();

        const renderTime = performance.now();
        let totalScanLineTime = 0;

        for (let j = 0; j < this.imageHeight; j++) {
            const scanLineStartTime = performance.now();

            for (let i = 0; i < this.imageWidth; i++) {
                let pixelColor = Vec3.zero;
                for (let sample = 0; sample < this.renderOptions.samplesPerPixel; sample++) {
                    const ray = this.getRay(i, j);
                    pixelColor = pixelColor.plus(this.rayColor(ray, this.renderOptions.maxDepth, world));
                }

                const intensity = new Interval(0.0, 0.999);
                const finalPixelColor = pixelColor
                    .times(1 / this.renderOptions.samplesPerPixel)
                    .map((n) => Math.floor(256 * intensity.clamp(Math.sqrt(n))));

                const index = (j * this.imageWidth + i) * 4;
                data[index + 0] = finalPixelColor.r;
                data[index + 1] = finalPixelColor.g;
                data[index + 2] = finalPixelColor.b;
                data[index + 3] = 255;

                const now = performance.now();
                if (now - lastYieldTime > BATCH_THRESHOLD_MS) {
                    await scheduler.yield();
                    lastYieldTime = now;
                }
            }

            const scanLineTime = performance.now() - scanLineStartTime;
            totalScanLineTime += scanLineTime;

            console.log(`Scan lines remaining: ${this.imageHeight - j}, took ${scanLineTime}ms`);
        }

        const averageScanLineTime = totalScanLineTime / this.imageHeight;
        console.log(`Done rendering! Took ${renderTime}ms, average scan line took ${averageScanLineTime}`);
    }

    private getRay(i: number, j: number) {
        const offset = this.sampleSquare();
        const pixelSample = this.upperLeftPixelLocation
            .plus(this.pixelDeltaU.times(i + offset.x))
            .plus(this.pixelDeltaV.times(j + offset.y));

        const origin = this.sceneOptions.defocusAngle <= 0 ? this.center : this.defocusDiskSample();
        const direction = pixelSample.minus(origin);
        const time = Interval.unit.random();

        return new Ray(origin, direction, time);
    }

    private sampleSquare() {
        const interval = new Interval(-0.5, 0.5);
        return new Vec3(interval.random(), interval.random(), 0);
    }

    private defocusDiskSample() {
        const point = Vec3.randomInUnitDisk();
        return this.center.plus(this.defocusDiskU.times(point.x)).plus(this.defocusDiskV.times(point.y));
    }

    private rayColor(ray: Ray, depth: number, world: Hittable): Vec3 {
        if (depth <= 0) {
            return Vec3.zero;
        }

        const hit = world.hit(ray, new Interval(0.001, Infinity));

        if (!hit) {
            if (this.sceneOptions.background) {
                return this.sceneOptions.background;
            }

            const unitDirection = ray.direction.unitVector;
            const a = 0.5 * (unitDirection.y + 1.0);
            return new Vec3(1.0, 1.0, 1.0).times(1.0 - a).plus(new Vec3(0.5, 0.7, 1.0).times(a));
        }

        const emissionColor = hit.material.emitted(hit.u, hit.v, hit.point);
        const materialResult = hit.material.scatter(ray, hit);
        if (!materialResult) {
            return emissionColor;
        }

        const scatterColor = materialResult.attenuation.times(
            this.rayColor(materialResult.scattered, depth - 1, world),
        );

        return emissionColor.plus(scatterColor);
    }
}
